{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Learning sgRNA predictors from empirical data\n",
    "    * Load scripts and empirical data\n",
    "    * Generate TSS annotation using FANTOM dataset\n",
    "    * Calculate parameters for empirical sgRNAs\n",
    "    * Fit parameters\n",
    "2. Applying machine learning model to predict sgRNA activity\n",
    "    * Find all sgRNAs in genomic regions of interest \n",
    "    * Predicting sgRNA activity\n",
    "3. Construct sgRNA libraries\n",
    "    * Score sgRNAs for off-target potential\n",
    "* Pick the top sgRNAs for a library, given predicted activity scores and off-target filtering\n",
    "* Design negative controls matching the base composition of the library\n",
    "* Finalizing library design"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Learning sgRNA predictors from empirical data\n",
    "## Load scripts and empirical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "#sys.path.insert(0, '../ScreenProcessing/')\n",
    "LARGE_FILE_DIR= '../../../../data/'\n",
    "\n",
    "FASTA_FILE_OF_GENOME= LARGE_FILE_DIR + 'large_data_files/hg19.fa'\n",
    "GTF_FILE_FROM_GENCODE = LARGE_FILE_DIR + 'large_data_files/gencode.v19.annotation.gtf'\n",
    "\n",
    "TSS_TABLE_PATH='data_files/human_tssTable.txt'\n",
    "P1P2_TABLE_PATH='data_files/human_p1p2Table.txt'\n",
    "\n",
    "DISCRIMANT_THRESHOLD_eg20=20\n",
    "FANTOM_TSS_ANNOTATION_BED= LARGE_FILE_DIR + 'large_data_files/TSS_human.sorted.bed.gz'\n",
    "HGNC_SYMBOL_LOOKUP_TABLE= LARGE_FILE_DIR + 'large_data_files/hgnc_complete_set_2020-08-01.txt'\n",
    "\n",
    "\n",
    "LIBRARY_TABLE_PATH='data_files/CRISPRi_trainingdata_libraryTable.txt'\n",
    "SGINFO_TABLE_PATH='data_files/CRISPRi_trainingdata_sgRNAInfoTable.txt'\n",
    "\n",
    "\n",
    "IGEM_EXCEL_FILE= LARGE_FILE_DIR + 'large_data_files/Sam_Perli_qRT_pcr_data_per_gene.xlsx'\n",
    "\n",
    "\n",
    "PICKLE_FILE = 'igem_v1_esitmator'\n",
    "TRANSFORMED_PARAM_HEADER='igem_v1_transformed_param_header'\n",
    "PREDICTED_SCORE_TABLE='igem_v1_predicted_score_table'\n",
    "TEMP_FASTQ_FILE='igem_v1_temp_fastq_file'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#TSS_TABLE_BASED_ON_ENSEMBL ='data_files/human_tssTable.txt\n",
    "#FANTOM_TSS_ANNOTATION_BED\n",
    "\n",
    "\n",
    "%run sgRNA_learning.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>protospacer sequence</th>\n",
       "      <th>selection rank</th>\n",
       "      <th>predicted score</th>\n",
       "      <th>empirical score</th>\n",
       "      <th>off-target stringency</th>\n",
       "      <th>Sublibrary half</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>Measured by Dr. Perli</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>[gene_strandtargeted_PAMcoordinate.sgRNAlength-transcript]</th>\n",
       "      <td>[gene targeted by the sgRNA, or \"negative_cont...</td>\n",
       "      <td>[TSS targeted by the sgRNA; see Table S2]</td>\n",
       "      <td>[protospacer sequence; 5'G is included whether...</td>\n",
       "      <td>[for each gene/transcript pair, the order in w...</td>\n",
       "      <td>[predicted score by CRISPRi-v2.1 algorithm]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[stringency level at which sgRNA passed thresh...</td>\n",
       "      <td>[Top5 sgRNAs per TSS or Supp5 sgRNAs to create...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830441.23-P1</th>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>P1</td>\n",
       "      <td>GAGTCGGAGCTCTATGGAGG</td>\n",
       "      <td>1</td>\n",
       "      <td>1.10756</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Top5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830256.23-P1</th>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>P1</td>\n",
       "      <td>GGTGGCAGCTGCTGAGTTCT</td>\n",
       "      <td>2</td>\n",
       "      <td>0.755364</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Top5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.7060</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830444.23-P1</th>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>P1</td>\n",
       "      <td>GGTGAGTCGGAGCTCTATGG</td>\n",
       "      <td>3</td>\n",
       "      <td>1.16826</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Top5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830447.23-P1</th>\n",
       "      <td>EIF4G2</td>\n",
       "      <td>P1</td>\n",
       "      <td>GGCAGTGAGTCGGAGCTCTA</td>\n",
       "      <td>4</td>\n",
       "      <td>1.10644</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Top5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2769</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC6_+_88969345.23-P1P2</th>\n",
       "      <td>ZCCHC6</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>GGTGGGGCCGAAGGAAAACA</td>\n",
       "      <td>6</td>\n",
       "      <td>0.773068</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Supp5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.4000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC6_-_88969249.23-P1P2</th>\n",
       "      <td>ZCCHC6</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>GAACTGGAGAATCAGCCCGG</td>\n",
       "      <td>7</td>\n",
       "      <td>0.765829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Supp5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC6_+_88969297.23-P1P2</th>\n",
       "      <td>ZCCHC6</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>GAGGCTGGGGACCGCGGCGA</td>\n",
       "      <td>8</td>\n",
       "      <td>0.735422</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Supp5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC6_-_88969214.23-P1P2</th>\n",
       "      <td>ZCCHC6</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>GCCAGGCCGGACAGCTACTC</td>\n",
       "      <td>9</td>\n",
       "      <td>0.517569</td>\n",
       "      <td>0.903625</td>\n",
       "      <td>0</td>\n",
       "      <td>Supp5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC6_-_88969308.23-P1P2</th>\n",
       "      <td>ZCCHC6</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>GGCACTCACCACCTTCGCCG</td>\n",
       "      <td>10</td>\n",
       "      <td>0.70495</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>Supp5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>191 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                 gene  \\\n",
       "sgID                                                                                                    \n",
       "[gene_strandtargeted_PAMcoordinate.sgRNAlength-...  [gene targeted by the sgRNA, or \"negative_cont...   \n",
       "EIF4G2_+_10830441.23-P1                                                                        EIF4G2   \n",
       "EIF4G2_+_10830256.23-P1                                                                        EIF4G2   \n",
       "EIF4G2_+_10830444.23-P1                                                                        EIF4G2   \n",
       "EIF4G2_+_10830447.23-P1                                                                        EIF4G2   \n",
       "...                                                                                               ...   \n",
       "ZCCHC6_+_88969345.23-P1P2                                                                      ZCCHC6   \n",
       "ZCCHC6_-_88969249.23-P1P2                                                                      ZCCHC6   \n",
       "ZCCHC6_+_88969297.23-P1P2                                                                      ZCCHC6   \n",
       "ZCCHC6_-_88969214.23-P1P2                                                                      ZCCHC6   \n",
       "ZCCHC6_-_88969308.23-P1P2                                                                      ZCCHC6   \n",
       "\n",
       "                                                                                   transcript  \\\n",
       "sgID                                                                                            \n",
       "[gene_strandtargeted_PAMcoordinate.sgRNAlength-...  [TSS targeted by the sgRNA; see Table S2]   \n",
       "EIF4G2_+_10830441.23-P1                                                                    P1   \n",
       "EIF4G2_+_10830256.23-P1                                                                    P1   \n",
       "EIF4G2_+_10830444.23-P1                                                                    P1   \n",
       "EIF4G2_+_10830447.23-P1                                                                    P1   \n",
       "...                                                                                       ...   \n",
       "ZCCHC6_+_88969345.23-P1P2                                                                P1P2   \n",
       "ZCCHC6_-_88969249.23-P1P2                                                                P1P2   \n",
       "ZCCHC6_+_88969297.23-P1P2                                                                P1P2   \n",
       "ZCCHC6_-_88969214.23-P1P2                                                                P1P2   \n",
       "ZCCHC6_-_88969308.23-P1P2                                                                P1P2   \n",
       "\n",
       "                                                                                 protospacer sequence  \\\n",
       "sgID                                                                                                    \n",
       "[gene_strandtargeted_PAMcoordinate.sgRNAlength-...  [protospacer sequence; 5'G is included whether...   \n",
       "EIF4G2_+_10830441.23-P1                                                          GAGTCGGAGCTCTATGGAGG   \n",
       "EIF4G2_+_10830256.23-P1                                                          GGTGGCAGCTGCTGAGTTCT   \n",
       "EIF4G2_+_10830444.23-P1                                                          GGTGAGTCGGAGCTCTATGG   \n",
       "EIF4G2_+_10830447.23-P1                                                          GGCAGTGAGTCGGAGCTCTA   \n",
       "...                                                                                               ...   \n",
       "ZCCHC6_+_88969345.23-P1P2                                                        GGTGGGGCCGAAGGAAAACA   \n",
       "ZCCHC6_-_88969249.23-P1P2                                                        GAACTGGAGAATCAGCCCGG   \n",
       "ZCCHC6_+_88969297.23-P1P2                                                        GAGGCTGGGGACCGCGGCGA   \n",
       "ZCCHC6_-_88969214.23-P1P2                                                        GCCAGGCCGGACAGCTACTC   \n",
       "ZCCHC6_-_88969308.23-P1P2                                                        GGCACTCACCACCTTCGCCG   \n",
       "\n",
       "                                                                                       selection rank  \\\n",
       "sgID                                                                                                    \n",
       "[gene_strandtargeted_PAMcoordinate.sgRNAlength-...  [for each gene/transcript pair, the order in w...   \n",
       "EIF4G2_+_10830441.23-P1                                                                             1   \n",
       "EIF4G2_+_10830256.23-P1                                                                             2   \n",
       "EIF4G2_+_10830444.23-P1                                                                             3   \n",
       "EIF4G2_+_10830447.23-P1                                                                             4   \n",
       "...                                                                                               ...   \n",
       "ZCCHC6_+_88969345.23-P1P2                                                                           6   \n",
       "ZCCHC6_-_88969249.23-P1P2                                                                           7   \n",
       "ZCCHC6_+_88969297.23-P1P2                                                                           8   \n",
       "ZCCHC6_-_88969214.23-P1P2                                                                           9   \n",
       "ZCCHC6_-_88969308.23-P1P2                                                                          10   \n",
       "\n",
       "                                                                                predicted score  \\\n",
       "sgID                                                                                              \n",
       "[gene_strandtargeted_PAMcoordinate.sgRNAlength-...  [predicted score by CRISPRi-v2.1 algorithm]   \n",
       "EIF4G2_+_10830441.23-P1                                                                 1.10756   \n",
       "EIF4G2_+_10830256.23-P1                                                                0.755364   \n",
       "EIF4G2_+_10830444.23-P1                                                                 1.16826   \n",
       "EIF4G2_+_10830447.23-P1                                                                 1.10644   \n",
       "...                                                                                         ...   \n",
       "ZCCHC6_+_88969345.23-P1P2                                                              0.773068   \n",
       "ZCCHC6_-_88969249.23-P1P2                                                              0.765829   \n",
       "ZCCHC6_+_88969297.23-P1P2                                                              0.735422   \n",
       "ZCCHC6_-_88969214.23-P1P2                                                              0.517569   \n",
       "ZCCHC6_-_88969308.23-P1P2                                                               0.70495   \n",
       "\n",
       "                                                    empirical score  \\\n",
       "sgID                                                                  \n",
       "[gene_strandtargeted_PAMcoordinate.sgRNAlength-...              NaN   \n",
       "EIF4G2_+_10830441.23-P1                                         NaN   \n",
       "EIF4G2_+_10830256.23-P1                                         NaN   \n",
       "EIF4G2_+_10830444.23-P1                                         NaN   \n",
       "EIF4G2_+_10830447.23-P1                                         NaN   \n",
       "...                                                             ...   \n",
       "ZCCHC6_+_88969345.23-P1P2                                       NaN   \n",
       "ZCCHC6_-_88969249.23-P1P2                                       NaN   \n",
       "ZCCHC6_+_88969297.23-P1P2                                       NaN   \n",
       "ZCCHC6_-_88969214.23-P1P2                                  0.903625   \n",
       "ZCCHC6_-_88969308.23-P1P2                                       NaN   \n",
       "\n",
       "                                                                                off-target stringency  \\\n",
       "sgID                                                                                                    \n",
       "[gene_strandtargeted_PAMcoordinate.sgRNAlength-...  [stringency level at which sgRNA passed thresh...   \n",
       "EIF4G2_+_10830441.23-P1                                                                             0   \n",
       "EIF4G2_+_10830256.23-P1                                                                             0   \n",
       "EIF4G2_+_10830444.23-P1                                                                             0   \n",
       "EIF4G2_+_10830447.23-P1                                                                             0   \n",
       "...                                                                                               ...   \n",
       "ZCCHC6_+_88969345.23-P1P2                                                                           0   \n",
       "ZCCHC6_-_88969249.23-P1P2                                                                           0   \n",
       "ZCCHC6_+_88969297.23-P1P2                                                                           0   \n",
       "ZCCHC6_-_88969214.23-P1P2                                                                           0   \n",
       "ZCCHC6_-_88969308.23-P1P2                                                                           0   \n",
       "\n",
       "                                                                                      Sublibrary half  \\\n",
       "sgID                                                                                                    \n",
       "[gene_strandtargeted_PAMcoordinate.sgRNAlength-...  [Top5 sgRNAs per TSS or Supp5 sgRNAs to create...   \n",
       "EIF4G2_+_10830441.23-P1                                                                          Top5   \n",
       "EIF4G2_+_10830256.23-P1                                                                          Top5   \n",
       "EIF4G2_+_10830444.23-P1                                                                          Top5   \n",
       "EIF4G2_+_10830447.23-P1                                                                          Top5   \n",
       "...                                                                                               ...   \n",
       "ZCCHC6_+_88969345.23-P1P2                                                                       Supp5   \n",
       "ZCCHC6_-_88969249.23-P1P2                                                                       Supp5   \n",
       "ZCCHC6_+_88969297.23-P1P2                                                                       Supp5   \n",
       "ZCCHC6_-_88969214.23-P1P2                                                                       Supp5   \n",
       "ZCCHC6_-_88969308.23-P1P2                                                                       Supp5   \n",
       "\n",
       "                                                   Unnamed: 9 Unnamed: 10  \\\n",
       "sgID                                                                        \n",
       "[gene_strandtargeted_PAMcoordinate.sgRNAlength-...        NaN         NaN   \n",
       "EIF4G2_+_10830441.23-P1                                   NaN         NaN   \n",
       "EIF4G2_+_10830256.23-P1                                   NaN         NaN   \n",
       "EIF4G2_+_10830444.23-P1                                   NaN         NaN   \n",
       "EIF4G2_+_10830447.23-P1                                   NaN         NaN   \n",
       "...                                                       ...         ...   \n",
       "ZCCHC6_+_88969345.23-P1P2                                 NaN         NaN   \n",
       "ZCCHC6_-_88969249.23-P1P2                                 NaN         NaN   \n",
       "ZCCHC6_+_88969297.23-P1P2                                 NaN         NaN   \n",
       "ZCCHC6_-_88969214.23-P1P2                                 NaN         NaN   \n",
       "ZCCHC6_-_88969308.23-P1P2                                 NaN         NaN   \n",
       "\n",
       "                                                    Measured by Dr. Perli  \n",
       "sgID                                                                       \n",
       "[gene_strandtargeted_PAMcoordinate.sgRNAlength-...                    NaN  \n",
       "EIF4G2_+_10830441.23-P1                                            0.1712  \n",
       "EIF4G2_+_10830256.23-P1                                            0.7060  \n",
       "EIF4G2_+_10830444.23-P1                                            0.0318  \n",
       "EIF4G2_+_10830447.23-P1                                            0.2769  \n",
       "...                                                                   ...  \n",
       "ZCCHC6_+_88969345.23-P1P2                                          0.4000  \n",
       "ZCCHC6_-_88969249.23-P1P2                                          0.0700  \n",
       "ZCCHC6_+_88969297.23-P1P2                                          0.2300  \n",
       "ZCCHC6_-_88969214.23-P1P2                                          0.1200  \n",
       "ZCCHC6_-_88969308.23-P1P2                                          0.1800  \n",
       "\n",
       "[191 rows x 11 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#pd.read_excel(IGEM_EXCEL_FILE, index_col=0) \n",
    "df=pd.read_excel(open(IGEM_EXCEL_FILE, 'rb'),\n",
    "              sheet_name='Sheet1')  \n",
    "df =df.dropna( subset = ['sgID'])\n",
    "df.set_index('sgID')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.columns\n",
    "libraryTable_research = df[['sgID','gene','transcript','protospacer sequence']]\n",
    "#libraryTable_research.rename(columns={'protospacer sequence':'sequence'})\n",
    "libraryTable_research= libraryTable_research.loc[1:,].set_index('sgID').rename(columns={'protospacer sequence':'sequence'})\n",
    "libraryTable_research\n",
    "\n",
    "#Select all genes except last one\n",
    "libraryTable_subset = libraryTable_research\n",
    "libraryTable_subset = libraryTable_research[ libraryTable_research['gene'] != 'EIF4G1' ]\n",
    "sgInfoTable = parseAllSgIds(libraryTable_subset)\n",
    "\n",
    "normedScores = df[['sgID','gene','Measured by Dr. Perli']].loc[1:,]\n",
    "normedScores = normedScores [ normedScores['gene'] != 'EIF4G1']\n",
    "\n",
    "#normedScores = normedScores [['sgID','Measured by Dr. Perli']].set_index('sgID').rename(columns={'Measured by Dr. Perli':''})\n",
    "#normedScores['Measured by Dr. Perli']= (-1)*normedScores['Measured by Dr. Perli']\n",
    "normedScores['Measured by Dr. Perli']= 1/normedScores['Measured by Dr. Perli']\n",
    "normedScores = normedScores [['sgID','Measured by Dr. Perli']].set_index('sgID').rename(columns={'Measured by Dr. Perli':''})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830441.23-P1</th>\n",
       "      <td>5.841121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830256.23-P1</th>\n",
       "      <td>1.416431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830444.23-P1</th>\n",
       "      <td>31.446541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830447.23-P1</th>\n",
       "      <td>3.611412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830421.23-P1</th>\n",
       "      <td>21.645022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC6_+_88969345.23-P1P2</th>\n",
       "      <td>2.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC6_-_88969249.23-P1P2</th>\n",
       "      <td>14.285714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC6_+_88969297.23-P1P2</th>\n",
       "      <td>4.347826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC6_-_88969214.23-P1P2</th>\n",
       "      <td>8.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCCHC6_-_88969308.23-P1P2</th>\n",
       "      <td>5.555556</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>180 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    \n",
       "sgID                                \n",
       "EIF4G2_+_10830441.23-P1     5.841121\n",
       "EIF4G2_+_10830256.23-P1     1.416431\n",
       "EIF4G2_+_10830444.23-P1    31.446541\n",
       "EIF4G2_+_10830447.23-P1     3.611412\n",
       "EIF4G2_+_10830421.23-P1    21.645022\n",
       "...                              ...\n",
       "ZCCHC6_+_88969345.23-P1P2   2.500000\n",
       "ZCCHC6_-_88969249.23-P1P2  14.285714\n",
       "ZCCHC6_+_88969297.23-P1P2   4.347826\n",
       "ZCCHC6_-_88969214.23-P1P2   8.333333\n",
       "ZCCHC6_-_88969308.23-P1P2   5.555556\n",
       "\n",
       "[180 rows x 1 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#print(normedScores.head())\n",
    "#normedScores.describe()\n",
    "normedScores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading genome file...Done\n",
      "Loading annotation file...Done\n"
     ]
    }
   ],
   "source": [
    "genomeDict = loadGenomeAsDict(FASTA_FILE_OF_GENOME)\n",
    "gencodeData = loadGencodeData(GTF_FILE_FROM_GENCODE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene          180\n",
      "transcript    180\n",
      "sequence      180\n",
      "dtype: int64\n",
      "Index(['gene', 'transcript', 'sequence'], dtype='object')\n",
      "                           gene transcript              sequence\n",
      "sgID                                                            \n",
      "EIF4G2_+_10830441.23-P1  EIF4G2         P1  GAGTCGGAGCTCTATGGAGG\n",
      "EIF4G2_+_10830256.23-P1  EIF4G2         P1  GGTGGCAGCTGCTGAGTTCT\n",
      "EIF4G2_+_10830444.23-P1  EIF4G2         P1  GGTGAGTCGGAGCTCTATGG\n",
      "EIF4G2_+_10830447.23-P1  EIF4G2         P1  GGCAGTGAGTCGGAGCTCTA\n",
      "EIF4G2_+_10830421.23-P1  EIF4G2         P1  GGGCAGCGGGTACCGAGTGG\n",
      "Sublibrary           0\n",
      "strand             180\n",
      "gene_name          180\n",
      "position           180\n",
      "length             180\n",
      "pass_score         180\n",
      "transcript_list    180\n",
      "pam coordinate     180\n",
      "dtype: int64\n",
      "Index(['Sublibrary', 'strand', 'gene_name', 'position', 'length', 'pass_score',\n",
      "       'transcript_list', 'pam coordinate'],\n",
      "      dtype='object')\n",
      "                        Sublibrary strand gene_name  position  length  \\\n",
      "sgID                                                                    \n",
      "EIF4G2_+_10830441.23-P1       None      +    EIF4G2  10830441      23   \n",
      "EIF4G2_+_10830256.23-P1       None      +    EIF4G2  10830256      23   \n",
      "EIF4G2_+_10830444.23-P1       None      +    EIF4G2  10830444      23   \n",
      "EIF4G2_+_10830447.23-P1       None      +    EIF4G2  10830447      23   \n",
      "EIF4G2_+_10830421.23-P1       None      +    EIF4G2  10830421      23   \n",
      "\n",
      "                        pass_score transcript_list  pam coordinate  \n",
      "sgID                                                                \n",
      "EIF4G2_+_10830441.23-P1         P1            [P1]        10830441  \n",
      "EIF4G2_+_10830256.23-P1         P1            [P1]        10830256  \n",
      "EIF4G2_+_10830444.23-P1         P1            [P1]        10830444  \n",
      "EIF4G2_+_10830447.23-P1         P1            [P1]        10830447  \n",
      "EIF4G2_+_10830421.23-P1         P1            [P1]        10830421  \n",
      "    180\n",
      "dtype: int64\n",
      "Index([''], dtype='object')\n",
      "                                  \n",
      "sgID                              \n",
      "EIF4G2_+_10830441.23-P1   5.841121\n",
      "EIF4G2_+_10830256.23-P1   1.416431\n",
      "EIF4G2_+_10830444.23-P1  31.446541\n",
      "EIF4G2_+_10830447.23-P1   3.611412\n",
      "EIF4G2_+_10830421.23-P1  21.645022\n"
     ]
    }
   ],
   "source": [
    "print (libraryTable_subset.count())\n",
    "print (libraryTable_subset.columns)\n",
    "print (libraryTable_subset.head())\n",
    "\n",
    "print (sgInfoTable.count())\n",
    "print (sgInfoTable.columns)\n",
    "print (sgInfoTable.head())\n",
    "#print (phenotypeTable.columns)\n",
    "#print (geneFoldList)\n",
    "\n",
    "print (normedScores.count())\n",
    "print (normedScores.columns)\n",
    "print (normedScores.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate TSS annotation using FANTOM dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>position</th>\n",
       "      <th>strand</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>cage peak ranges</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th>transcripts</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <th>all</th>\n",
       "      <td>58864864.0</td>\n",
       "      <td>-</td>\n",
       "      <td>chr19</td>\n",
       "      <td>[(58864822, 58864847), (58864848, 58864868)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">A1CF</th>\n",
       "      <th>ENST00000373993.1</th>\n",
       "      <td>52619744.0</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000374001.2,ENST00000373997.3,ENST00000282641.2</th>\n",
       "      <td>52645434.0</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10</td>\n",
       "      <td>[(52645379, 52645393), (52645416, 52645444)]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <th>all</th>\n",
       "      <td>9268752.0</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12</td>\n",
       "      <td>[(9268547, 9268556), (9268559, 9268568), (9268...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <th>all</th>\n",
       "      <td>8975067.0</td>\n",
       "      <td>+</td>\n",
       "      <td>chr12</td>\n",
       "      <td>[(8975061, 8975072), (8975101, 8975108), (8975...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                            position strand  \\\n",
       "gene  transcripts                                                             \n",
       "A1BG  all                                                 58864864.0      -   \n",
       "A1CF  ENST00000373993.1                                   52619744.0      -   \n",
       "      ENST00000374001.2,ENST00000373997.3,ENST0000028...  52645434.0      -   \n",
       "A2M   all                                                  9268752.0      -   \n",
       "A2ML1 all                                                  8975067.0      +   \n",
       "\n",
       "                                                         chromosome  \\\n",
       "gene  transcripts                                                     \n",
       "A1BG  all                                                     chr19   \n",
       "A1CF  ENST00000373993.1                                       chr10   \n",
       "      ENST00000374001.2,ENST00000373997.3,ENST0000028...      chr10   \n",
       "A2M   all                                                     chr12   \n",
       "A2ML1 all                                                     chr12   \n",
       "\n",
       "                                                                                           cage peak ranges  \n",
       "gene  transcripts                                                                                            \n",
       "A1BG  all                                                      [(58864822, 58864847), (58864848, 58864868)]  \n",
       "A1CF  ENST00000373993.1                                                                                  []  \n",
       "      ENST00000374001.2,ENST00000373997.3,ENST0000028...       [(52645379, 52645393), (52645416, 52645444)]  \n",
       "A2M   all                                                 [(9268547, 9268556), (9268559, 9268568), (9268...  \n",
       "A2ML1 all                                                 [(8975061, 8975072), (8975101, 8975108), (8975...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import ast\n",
    "#save tables for downstream use\n",
    "#tssTable = pd.read_csv(TSS_TABLE_PATH,sep='\\t', index_col=range(2))\n",
    "tssTable = pd.read_csv(TSS_TABLE_PATH,sep='\\t', index_col=[0,1])\n",
    "\n",
    "\n",
    "#p1p2Table = pd.read_csv(P1P2_TABLE_PATH,sep='\\t', header=0, index_col=range(2))\n",
    "p1p2Table = pd.read_csv(P1P2_TABLE_PATH,sep='\\t', header=0, index_col=[0,1], converters={\"primary TSS\": ast.literal_eval, \"secondary TSS\": ast.literal_eval})\n",
    "\n",
    "tssTable.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## Calculate parameters for empirical sgRNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Load bigwig files for any chromatin data of interest\n",
    "#bwhandleDict = {'dnase':BigWigFile(open('ENCODE_data/wgEncodeOpenChromDnaseK562BaseOverlapSignalV2.bigWig','rb')),\n",
    "#'faire':BigWigFile(open('ENCODE_data/wgEncodeOpenChromFaireK562Sig.bigWig','rb')),\n",
    "#'mnase':BigWigFile(open('ENCODE_data/wgEncodeSydhNsomeK562Sig.bigWig','rb'))}\n",
    "bwhandleDict = {'dnase':BigWigFile(open(LARGE_FILE_DIR + 'large_data_files/wgEncodeOpenChromDnaseK562BaseOverlapSignalV2.bigWig','rb')),\n",
    "'faire':BigWigFile(open(LARGE_FILE_DIR + 'large_data_files/wgEncodeOpenChromFaireK562Sig.bigWig','rb')),\n",
    "'mnase':BigWigFile(open(LARGE_FILE_DIR + 'large_data_files/wgEncodeSydhNsomeK562Sig.bigWig','rb'))}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".....Done!"
     ]
    }
   ],
   "source": [
    "#import pdb\n",
    "#pdb.set_trace()\n",
    "#paramTable_trainingGuides = generateTypicalParamTable(libraryTable_subset,sgInfoTable, tssTable, p1p2Table, genomeDict, bwhandleDict)\n",
    "paramTable_trainingGuides = generateTypicalParamTableEx(libraryTable_subset,sgInfoTable, tssTable, p1p2Table, genomeDict, bwhandleDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th colspan=\"4\" halign=\"left\">distance</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>primary TSS-Up</th>\n",
       "      <th>primary TSS-Down</th>\n",
       "      <th>secondary TSS-Up</th>\n",
       "      <th>secondary TSS-Down</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>180.0</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>180.000000</td>\n",
       "      <td>180.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>20.0</td>\n",
       "      <td>139.138889</td>\n",
       "      <td>83.250000</td>\n",
       "      <td>114.972222</td>\n",
       "      <td>85.25000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.0</td>\n",
       "      <td>135.339662</td>\n",
       "      <td>133.875615</td>\n",
       "      <td>153.952559</td>\n",
       "      <td>148.96139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>20.0</td>\n",
       "      <td>-152.000000</td>\n",
       "      <td>-199.000000</td>\n",
       "      <td>-230.000000</td>\n",
       "      <td>-252.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>20.0</td>\n",
       "      <td>48.750000</td>\n",
       "      <td>-5.000000</td>\n",
       "      <td>16.000000</td>\n",
       "      <td>-7.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>20.0</td>\n",
       "      <td>88.000000</td>\n",
       "      <td>37.000000</td>\n",
       "      <td>84.000000</td>\n",
       "      <td>56.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>20.0</td>\n",
       "      <td>220.000000</td>\n",
       "      <td>156.000000</td>\n",
       "      <td>203.500000</td>\n",
       "      <td>164.25000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>20.0</td>\n",
       "      <td>502.000000</td>\n",
       "      <td>472.000000</td>\n",
       "      <td>630.000000</td>\n",
       "      <td>551.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      length       distance                                    \\\n",
       "      length primary TSS-Up primary TSS-Down secondary TSS-Up   \n",
       "count  180.0     180.000000       180.000000       180.000000   \n",
       "mean    20.0     139.138889        83.250000       114.972222   \n",
       "std      0.0     135.339662       133.875615       153.952559   \n",
       "min     20.0    -152.000000      -199.000000      -230.000000   \n",
       "25%     20.0      48.750000        -5.000000        16.000000   \n",
       "50%     20.0      88.000000        37.000000        84.000000   \n",
       "75%     20.0     220.000000       156.000000       203.500000   \n",
       "max     20.0     502.000000       472.000000       630.000000   \n",
       "\n",
       "                          \n",
       "      secondary TSS-Down  \n",
       "count          180.00000  \n",
       "mean            85.25000  \n",
       "std            148.96139  \n",
       "min           -252.00000  \n",
       "25%             -7.00000  \n",
       "50%             56.00000  \n",
       "75%            164.25000  \n",
       "max            551.00000  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paramTable_trainingGuides.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th colspan=\"4\" halign=\"left\">distance</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>length</th>\n",
       "      <th>primary TSS-Up</th>\n",
       "      <th>primary TSS-Down</th>\n",
       "      <th>secondary TSS-Up</th>\n",
       "      <th>secondary TSS-Down</th>\n",
       "      <th>same strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830441.23-P1</th>\n",
       "      <td>20</td>\n",
       "      <td>53</td>\n",
       "      <td>22</td>\n",
       "      <td>53</td>\n",
       "      <td>22</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830256.23-P1</th>\n",
       "      <td>20</td>\n",
       "      <td>238</td>\n",
       "      <td>207</td>\n",
       "      <td>238</td>\n",
       "      <td>207</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830444.23-P1</th>\n",
       "      <td>20</td>\n",
       "      <td>50</td>\n",
       "      <td>19</td>\n",
       "      <td>50</td>\n",
       "      <td>19</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830447.23-P1</th>\n",
       "      <td>20</td>\n",
       "      <td>47</td>\n",
       "      <td>16</td>\n",
       "      <td>47</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G2_+_10830421.23-P1</th>\n",
       "      <td>20</td>\n",
       "      <td>73</td>\n",
       "      <td>42</td>\n",
       "      <td>73</td>\n",
       "      <td>42</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        length       distance                   \\\n",
       "                        length primary TSS-Up primary TSS-Down   \n",
       "EIF4G2_+_10830441.23-P1     20             53               22   \n",
       "EIF4G2_+_10830256.23-P1     20            238              207   \n",
       "EIF4G2_+_10830444.23-P1     20             50               19   \n",
       "EIF4G2_+_10830447.23-P1     20             47               16   \n",
       "EIF4G2_+_10830421.23-P1     20             73               42   \n",
       "\n",
       "                                                                 strand  \n",
       "                        secondary TSS-Up secondary TSS-Down same strand  \n",
       "EIF4G2_+_10830441.23-P1               53                 22       False  \n",
       "EIF4G2_+_10830256.23-P1              238                207       False  \n",
       "EIF4G2_+_10830444.23-P1               50                 19       False  \n",
       "EIF4G2_+_10830447.23-P1               47                 16       False  \n",
       "EIF4G2_+_10830421.23-P1               73                 42       False  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print( len(paramTable_trainingGuides.columns) )\n",
    "paramTable_trainingGuides.columns\n",
    "paramTable_trainingGuides.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fit parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#populate table of fitting parameters\n",
    "typeList = ['binnable_onehot', \n",
    "            'continuous', 'continuous', 'continuous', 'continuous',\n",
    "#            'binnable_onehot','binnable_onehot','binnable_onehot','binnable_onehot',\n",
    "#            'binnable_onehot','binnable_onehot','binnable_onehot','binnable_onehot','binnable_onehot','binnable_onehot','binnable_onehot',\n",
    "            'binary']\n",
    "#typeList.extend(['binary']*160)\n",
    "#typeList.extend(['binary']*(16*38))\n",
    "#typeList.extend(['binnable_onehot']*3)\n",
    "#typeList.extend(['binnable_onehot']*2)\n",
    "#typeList.extend(['binary']*18)\n",
    "fitTable = pd.DataFrame(typeList, index=paramTable_trainingGuides.columns, columns=['type'])\n",
    "#we may have to reduce min edge data from 50\n",
    "MIN_EDGE_DATA=15\n",
    "fitparams =[{'bin width':1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "            {'C':[.01,.05, .1,.5], 'gamma':[.000001, .00005,.0001,.0005]},\n",
    "            {'C':[.01,.05, .1,.5], 'gamma':[.000001, .00005,.0001,.0005]},\n",
    "            {'C':[.01,.05, .1,.5], 'gamma':[.000001, .00005,.0001,.0005]},\n",
    "            {'C':[.01,.05, .1,.5], 'gamma':[.000001, .00005,.0001,.0005]},\n",
    "#            {'bin width':1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':.1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':.1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':.1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':.1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':.1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':.1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':.1, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "            dict()]\n",
    "#fitparams.extend([dict()]*160)\n",
    "#fitparams.extend([dict()]*(16*38))\n",
    "#fitparams.extend([\n",
    "#            {'bin width':.15, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':.15, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':.15, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median}])\n",
    "#fitparams.extend([\n",
    "#            {'bin width':2, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median},\n",
    "#            {'bin width':2, 'min edge data':MIN_EDGE_DATA, 'bin function':np.median}])\n",
    "#fitparams.extend([dict()]*18)\n",
    "fitTable['params'] = fitparams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MultiIndex([(  'length',             'length'),\n",
       "            ('distance',     'primary TSS-Up'),\n",
       "            ('distance',   'primary TSS-Down'),\n",
       "            ('distance',   'secondary TSS-Up'),\n",
       "            ('distance', 'secondary TSS-Down'),\n",
       "            (  'strand',        'same strand')],\n",
       "           )"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "paramTable_trainingGuides.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import pdb\n",
    "#pdb.set_trace()\n",
    "#Train set includes all genes except 1 May have to retry this step and training multiple times until we get an estimator with atleast one feature.\n",
    "# \"Number of features used: \" printed at the end of below training cell should a value greather than zero.\n",
    "geneFoldList = getGeneFoldsEx(libraryTable_subset, 2, transcripts=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[([80,\n",
       "   81,\n",
       "   82,\n",
       "   83,\n",
       "   84,\n",
       "   85,\n",
       "   86,\n",
       "   87,\n",
       "   88,\n",
       "   89,\n",
       "   0,\n",
       "   1,\n",
       "   2,\n",
       "   3,\n",
       "   4,\n",
       "   5,\n",
       "   6,\n",
       "   7,\n",
       "   8,\n",
       "   9,\n",
       "   110,\n",
       "   111,\n",
       "   112,\n",
       "   113,\n",
       "   114,\n",
       "   115,\n",
       "   116,\n",
       "   117,\n",
       "   118,\n",
       "   119,\n",
       "   120,\n",
       "   121,\n",
       "   122,\n",
       "   123,\n",
       "   124,\n",
       "   125,\n",
       "   126,\n",
       "   127,\n",
       "   128,\n",
       "   129,\n",
       "   60,\n",
       "   61,\n",
       "   62,\n",
       "   63,\n",
       "   64,\n",
       "   65,\n",
       "   66,\n",
       "   67,\n",
       "   68,\n",
       "   69,\n",
       "   40,\n",
       "   41,\n",
       "   42,\n",
       "   43,\n",
       "   44,\n",
       "   45,\n",
       "   46,\n",
       "   47,\n",
       "   48,\n",
       "   49,\n",
       "   50,\n",
       "   51,\n",
       "   52,\n",
       "   53,\n",
       "   54,\n",
       "   55,\n",
       "   56,\n",
       "   57,\n",
       "   58,\n",
       "   59,\n",
       "   90,\n",
       "   91,\n",
       "   92,\n",
       "   93,\n",
       "   94,\n",
       "   95,\n",
       "   96,\n",
       "   97,\n",
       "   98,\n",
       "   99,\n",
       "   70,\n",
       "   71,\n",
       "   72,\n",
       "   73,\n",
       "   74,\n",
       "   75,\n",
       "   76,\n",
       "   77,\n",
       "   78,\n",
       "   79,\n",
       "   160,\n",
       "   161,\n",
       "   162,\n",
       "   163,\n",
       "   164,\n",
       "   165,\n",
       "   166,\n",
       "   167,\n",
       "   168,\n",
       "   169,\n",
       "   10,\n",
       "   11,\n",
       "   12,\n",
       "   13,\n",
       "   14,\n",
       "   15,\n",
       "   16,\n",
       "   17,\n",
       "   18,\n",
       "   19,\n",
       "   20,\n",
       "   21,\n",
       "   22,\n",
       "   23,\n",
       "   24,\n",
       "   25,\n",
       "   26,\n",
       "   27,\n",
       "   28,\n",
       "   29,\n",
       "   100,\n",
       "   101,\n",
       "   102,\n",
       "   103,\n",
       "   104,\n",
       "   105,\n",
       "   106,\n",
       "   107,\n",
       "   108,\n",
       "   109,\n",
       "   130,\n",
       "   131,\n",
       "   132,\n",
       "   133,\n",
       "   134,\n",
       "   135,\n",
       "   136,\n",
       "   137,\n",
       "   138,\n",
       "   139,\n",
       "   140,\n",
       "   141,\n",
       "   142,\n",
       "   143,\n",
       "   144,\n",
       "   145,\n",
       "   146,\n",
       "   147,\n",
       "   148,\n",
       "   149,\n",
       "   150,\n",
       "   151,\n",
       "   152,\n",
       "   153,\n",
       "   154,\n",
       "   155,\n",
       "   156,\n",
       "   157,\n",
       "   158,\n",
       "   159,\n",
       "   170,\n",
       "   171,\n",
       "   172,\n",
       "   173,\n",
       "   174,\n",
       "   175,\n",
       "   176,\n",
       "   177,\n",
       "   178,\n",
       "   179],\n",
       "  [30, 31, 32, 33, 34, 35, 36, 37, 38, 39]),\n",
       " ([80,\n",
       "   81,\n",
       "   82,\n",
       "   83,\n",
       "   84,\n",
       "   85,\n",
       "   86,\n",
       "   87,\n",
       "   88,\n",
       "   89,\n",
       "   0,\n",
       "   1,\n",
       "   2,\n",
       "   3,\n",
       "   4,\n",
       "   5,\n",
       "   6,\n",
       "   7,\n",
       "   8,\n",
       "   9,\n",
       "   110,\n",
       "   111,\n",
       "   112,\n",
       "   113,\n",
       "   114,\n",
       "   115,\n",
       "   116,\n",
       "   117,\n",
       "   118,\n",
       "   119,\n",
       "   120,\n",
       "   121,\n",
       "   122,\n",
       "   123,\n",
       "   124,\n",
       "   125,\n",
       "   126,\n",
       "   127,\n",
       "   128,\n",
       "   129,\n",
       "   60,\n",
       "   61,\n",
       "   62,\n",
       "   63,\n",
       "   64,\n",
       "   65,\n",
       "   66,\n",
       "   67,\n",
       "   68,\n",
       "   69,\n",
       "   40,\n",
       "   41,\n",
       "   42,\n",
       "   43,\n",
       "   44,\n",
       "   45,\n",
       "   46,\n",
       "   47,\n",
       "   48,\n",
       "   49,\n",
       "   50,\n",
       "   51,\n",
       "   52,\n",
       "   53,\n",
       "   54,\n",
       "   55,\n",
       "   56,\n",
       "   57,\n",
       "   58,\n",
       "   59,\n",
       "   90,\n",
       "   91,\n",
       "   92,\n",
       "   93,\n",
       "   94,\n",
       "   95,\n",
       "   96,\n",
       "   97,\n",
       "   98,\n",
       "   99,\n",
       "   70,\n",
       "   71,\n",
       "   72,\n",
       "   73,\n",
       "   74,\n",
       "   75,\n",
       "   76,\n",
       "   77,\n",
       "   78,\n",
       "   79,\n",
       "   160,\n",
       "   161,\n",
       "   162,\n",
       "   163,\n",
       "   164,\n",
       "   165,\n",
       "   166,\n",
       "   167,\n",
       "   168,\n",
       "   169,\n",
       "   20,\n",
       "   21,\n",
       "   22,\n",
       "   23,\n",
       "   24,\n",
       "   25,\n",
       "   26,\n",
       "   27,\n",
       "   28,\n",
       "   29,\n",
       "   30,\n",
       "   31,\n",
       "   32,\n",
       "   33,\n",
       "   34,\n",
       "   35,\n",
       "   36,\n",
       "   37,\n",
       "   38,\n",
       "   39,\n",
       "   100,\n",
       "   101,\n",
       "   102,\n",
       "   103,\n",
       "   104,\n",
       "   105,\n",
       "   106,\n",
       "   107,\n",
       "   108,\n",
       "   109,\n",
       "   130,\n",
       "   131,\n",
       "   132,\n",
       "   133,\n",
       "   134,\n",
       "   135,\n",
       "   136,\n",
       "   137,\n",
       "   138,\n",
       "   139,\n",
       "   140,\n",
       "   141,\n",
       "   142,\n",
       "   143,\n",
       "   144,\n",
       "   145,\n",
       "   146,\n",
       "   147,\n",
       "   148,\n",
       "   149,\n",
       "   150,\n",
       "   151,\n",
       "   152,\n",
       "   153,\n",
       "   154,\n",
       "   155,\n",
       "   156,\n",
       "   157,\n",
       "   158,\n",
       "   159,\n",
       "   170,\n",
       "   171,\n",
       "   172,\n",
       "   173,\n",
       "   174,\n",
       "   175,\n",
       "   176,\n",
       "   177,\n",
       "   178,\n",
       "   179],\n",
       "  [10, 11, 12, 13, 14, 15, 16, 17, 18, 19])]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneFoldList"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('distance', 'primary TSS-Up') {'C': 0.5, 'gamma': 1e-06}\n",
      "('distance', 'primary TSS-Down') {'C': 0.5, 'gamma': 1e-06}\n",
      "('distance', 'secondary TSS-Up') {'C': 0.01, 'gamma': 1e-06}\n",
      "('distance', 'secondary TSS-Down') {'C': 0.01, 'gamma': 1e-06}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/.local/lib/python3.8/site-packages/sklearn/utils/validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
      "  return f(**kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Prediction R^2: -300.048226814354\n",
      "Regression parameters: 0.5 2.6214991353519954\n",
      "Number of features used: 0\n",
      "('distance', 'primary TSS-Up') {'C': 0.05, 'gamma': 0.0005}\n",
      "('distance', 'primary TSS-Down') {'C': 0.5, 'gamma': 0.0001}\n",
      "('distance', 'secondary TSS-Up') {'C': 0.01, 'gamma': 0.0005}\n",
      "('distance', 'secondary TSS-Down') {'C': 0.01, 'gamma': 1e-06}\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/.local/lib/python3.8/site-packages/sklearn/utils/validation.py:72: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().\n",
      "  return f(**kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Prediction R^2: -3.254951545213915\n",
      "Regression parameters: 0.5 2.733062826146315\n",
      "Number of features used: 0\n"
     ]
    }
   ],
   "source": [
    "#import pdb\n",
    "#pdb.set_trace()\n",
    "\n",
    "#for each fold, fit parameters to training folds and measure ROC on test fold\n",
    "coefs = []\n",
    "scoreTups = []\n",
    "transformedParamTups = []\n",
    "\n",
    "for geneFold_train, geneFold_test in geneFoldList:\n",
    "\n",
    "    transformedParams_train, estimators = fitParams(paramTable_trainingGuides.loc[normedScores.dropna().index].iloc[geneFold_train], normedScores.loc[normedScores.dropna().index].iloc[geneFold_train], fitTable)\n",
    "\n",
    "    transformedParams_test = transformParams(paramTable_trainingGuides.loc[normedScores.dropna().index].iloc[geneFold_test], fitTable, estimators)\n",
    "    \n",
    "    reg = linear_model.ElasticNetCV(l1_ratio=[.5, .75, .9, .99,1], n_jobs=16, max_iter=2000)\n",
    "    \n",
    "    scaler = preprocessing.StandardScaler()\n",
    "    reg.fit(scaler.fit_transform(transformedParams_train), normedScores.loc[normedScores.dropna().index].iloc[geneFold_train])\n",
    "    predictedScores = pd.Series(reg.predict(scaler.transform(transformedParams_test)), index=transformedParams_test.index)\n",
    "    testScores = normedScores.loc[normedScores.dropna().index].iloc[geneFold_test]\n",
    "    \n",
    "    transformedParamTups.append((scaler.transform(transformedParams_train),scaler.transform(transformedParams_test)))\n",
    "    scoreTups.append((testScores, predictedScores))\n",
    "\n",
    "#  for now ignore one class errors.....    \n",
    "#    print ('Prediction AUC-ROC:', metrics.roc_auc_score((testScores >= .75).values, np.array(predictedScores.values,dtype='float64')))\n",
    "    print ('Prediction R^2:', reg.score(scaler.transform(transformedParams_test), testScores))\n",
    "    print ('Regression parameters:', reg.l1_ratio_, reg.alpha_)\n",
    "    coefs.append(pd.DataFrame(zip(*[abs(reg.coef_),reg.coef_]), index = transformedParams_test.columns, columns=['abs','true']))\n",
    "    print ('Number of features used:', len(coefs[-1]) - sum(coefs[-1]['abs'] < .00000000001))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#can select an arbitrary fold (as shown here simply the last one tested) to save state for reproducing estimators later\n",
    "#the pickling of the scikit-learn estimators/regressors will allow the model to be reloaded for prediction of other guide designs, \n",
    "#   but will not be compatible across scikit-learn versions, so it is important to preserve the training data and training/test folds\n",
    "#import cPickle\n",
    "import _pickle as cPickle\n",
    "\n",
    "estimatorString = cPickle.dumps((fitTable, estimators, scaler, reg, (geneFold_train, geneFold_test)))\n",
    "with open(PICKLE_FILE,'wb') as outfile:\n",
    "    outfile.write(estimatorString)\n",
    "    \n",
    "#also save the transformed parameters as these can slightly differ based on the automated binning strategy\n",
    "\n",
    "#transformedParams_train.head().to_csv(TRANSFORMED_PARAM_HEADER,sep='\\t')\n",
    "#iGEM  instaed of csv write the head in binary format so that it can be read properly. There were \n",
    "\n",
    "transformedParamsTrainHead = cPickle.dumps(transformedParams_train.head())\n",
    "\n",
    "with open(TRANSFORMED_PARAM_HEADER,'wb') as paramfile:\n",
    "    paramfile.write(transformedParamsTrainHead)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# 2. Applying machine learning model to predict sgRNA activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading genome file...Done\n"
     ]
    }
   ],
   "source": [
    "#starting from a new session for demonstration purposes:\n",
    "%run sgRNA_learning.py\n",
    "import _pickle as cPickle\n",
    "#import cPickle\n",
    "\n",
    "#load tssTable, p1p2Table, genome sequence, chromatin data\n",
    "tssTable = pd.read_csv(TSS_TABLE_PATH,sep='\\t', index_col=[0,1])\n",
    "\n",
    "p1p2Table = pd.read_csv(P1P2_TABLE_PATH,sep='\\t', header=0, index_col=[0,1])\n",
    "p1p2Table['primary TSS'] = p1p2Table['primary TSS'].apply(lambda tupString: (int(float(tupString.strip('()').split(', ')[0])), int(float(tupString.strip('()').split(', ')[1]))))\n",
    "p1p2Table['secondary TSS'] = p1p2Table['secondary TSS'].apply(lambda tupString: (int(float(tupString.strip('()').split(', ')[0])),int(float(tupString.strip('()').split(', ')[1]))))\n",
    "\n",
    "genomeDict = loadGenomeAsDict(FASTA_FILE_OF_GENOME)\n",
    "\n",
    "bwhandleDict = {'dnase':BigWigFile(open(LARGE_FILE_DIR + 'large_data_files/wgEncodeOpenChromDnaseK562BaseOverlapSignalV2.bigWig','rb')),\n",
    "'faire':BigWigFile(open(LARGE_FILE_DIR + 'large_data_files/wgEncodeOpenChromFaireK562Sig.bigWig','rb')),\n",
    "'mnase':BigWigFile(open(LARGE_FILE_DIR + 'large_data_files/wgEncodeSydhNsomeK562Sig.bigWig','rb'))}\n",
    "\n",
    "#load sgRNA prediction model saved after the parameter fitting step\n",
    "with open(PICKLE_FILE,'rb') as infile:\n",
    "    fitTable, estimators, scaler, reg, (geneFold_train, geneFold_test) = cPickle.load(infile)\n",
    "    \n",
    "#transformedParamHeader = pd.read_csv(TRANSFORMED_PARAM_HEADER,sep='\\t')\n",
    "\n",
    "#iGEM read the binary file\n",
    "#transformedParamHeader = pd.read_csv(TRANSFORMED_PARAM_HEADER,sep='\\t')\n",
    "\n",
    "with open(TRANSFORMED_PARAM_HEADER,'rb') as paraminfile:\n",
    "    transformedParamHeader = cPickle.load(paraminfile)\n",
    "\n",
    "transformedParams_train = transformedParamHeader"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Find all sgRNAs in genomic regions of interest "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#For IGEM for now try to score the training data itself.\n",
    "#libraryTable_new = libraryTable_subset\n",
    "#sgInfoTable_new = sgInfoTable\n",
    "\n",
    "#Select excluded gene fron training set\n",
    "libraryTable_new = libraryTable_research\n",
    "libraryTable_new = libraryTable_research[ libraryTable_research['gene'] == 'EIF4G1' ]\n",
    "\n",
    "#libraryTable_subset = libraryTable_research\n",
    "sgInfoTable_new = parseAllSgIds(libraryTable_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sgID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>EIF4G1_-_184032386.23-P1P2</th>\n",
       "      <td>EIF4G1</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>GCCGCCGAGCGGGATCTGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G1_-_184032400.23-P1P2</th>\n",
       "      <td>EIF4G1</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>GCTGTGCGGGGAGCCGGAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G1_-_184032364.23-P1P2</th>\n",
       "      <td>EIF4G1</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>GTGCGCCTGCGGAGAAGCGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G1_-_184032431.23-P1P2</th>\n",
       "      <td>EIF4G1</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>GACGTCTGTGCGGCTGCGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EIF4G1_-_184032787.23-P1P2</th>\n",
       "      <td>EIF4G1</td>\n",
       "      <td>P1P2</td>\n",
       "      <td>GCCCACGGGCGCAGGCACCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              gene transcript              sequence\n",
       "sgID                                                               \n",
       "EIF4G1_-_184032386.23-P1P2  EIF4G1       P1P2  GCCGCCGAGCGGGATCTGTG\n",
       "EIF4G1_-_184032400.23-P1P2  EIF4G1       P1P2  GCTGTGCGGGGAGCCGGAAA\n",
       "EIF4G1_-_184032364.23-P1P2  EIF4G1       P1P2  GTGCGCCTGCGGAGAAGCGG\n",
       "EIF4G1_-_184032431.23-P1P2  EIF4G1       P1P2  GACGTCTGTGCGGCTGCGTG\n",
       "EIF4G1_-_184032787.23-P1P2  EIF4G1       P1P2  GCCCACGGGCGCAGGCACCA"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "libraryTable_new.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Predicting sgRNA activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".....Done!"
     ]
    }
   ],
   "source": [
    "#calculate parameters for new sgRNAs\n",
    "#paramTable_new = generateTypicalParamTable(libraryTable_new, sgInfoTable_new, tssTable, p1p2Table, genomeDict, bwhandleDict)\n",
    "paramTable_new = generateTypicalParamTableEx(libraryTable_new, sgInfoTable_new, tssTable, p1p2Table, genomeDict, bwhandleDict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import pdb\n",
    "#pdb.set_trace()\n",
    "#transform and predict scores according to sgRNA prediction model\n",
    "transformedParams_new = transformParams(paramTable_new, fitTable, estimators)\n",
    "\n",
    "#reconcile any differences in column headers generated by automated binning\n",
    "colTups = []\n",
    "for (l1, l2), col in transformedParams_new.iteritems():\n",
    "    colTups.append((l1,str(l2)))\n",
    "transformedParams_new.columns = pd.MultiIndex.from_tuples(colTups)\n",
    "#iGEM in python 3 .loc with missing column headers is giving issues. So changing it to reindex # can use transformedParams_train if running sequentially otherwise use transformedParamHeader after running above step\n",
    "#predictedScores_new = pd.Series(reg.predict(scaler.transform(transformedParams_new.loc[:, transformedParamHeader.columns].fillna(0).values)), index=transformedParams_new.index)\n",
    "#predictedScores_new = pd.Series(reg.predict(scaler.transform(transformedParams_new.reindex(columns=transformedParamHeader.columns).fillna(0).values)), index=transformedParams_new.index)\n",
    "predictedScores_new = pd.Series(reg.predict(scaler.transform(transformedParams_new.reindex(columns=transformedParams_train.columns).fillna(0).values)), index=transformedParams_new.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EIF4G1_-_184032386.23-P1P2    6.480369\n",
       "EIF4G1_-_184032400.23-P1P2    6.480369\n",
       "EIF4G1_-_184032364.23-P1P2    6.480369\n",
       "EIF4G1_-_184032431.23-P1P2    6.480369\n",
       "EIF4G1_-_184032787.23-P1P2    6.480369\n",
       "dtype: float64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictedScores_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictedScores_new.to_csv(PREDICTED_SCORE_TABLE, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Construct sgRNA libraries\n",
    "## Score sgRNAs for off-target potential"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#There are many ways to score sgRNAs as off-target; below is one listed one method that is simple and flexible,\n",
    "#but ignores gapped alignments, alternate PAMs, and uses bowtie which may not be maximally sensitive in all cases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#iGEM the sequence length can be greather 22. So just added more pluses and adjusteed its length equal to sequence length.\n",
    "#iGEM Revisit to fix phred length properly.\n",
    "\n",
    "#output all sequences to a temporary FASTQ file for running bowtie alignment\n",
    "def outputTempBowtieFastq(libraryTable, outputFileName):\n",
    "    phredString = 'I4!=======44444++++++++++++++++' #weighting for how impactful mismatches are along sgRNA sequence \n",
    "    with open(outputFileName,'w') as outfile:\n",
    "        for name, row in libraryTable.iterrows():\n",
    "            outfile.write('@' + name + '\\n')\n",
    "            outfile.write('CCN' + str(Seq.Seq(row['sequence'][1:]).reverse_complement()) + '\\n')\n",
    "            outfile.write('+\\n')\n",
    "            outfile.write(phredString[0:3+len(str(Seq.Seq(row['sequence'][1:]).reverse_complement()))] + '\\n')\n",
    "            \n",
    "outputTempBowtieFastq(libraryTable_new, TEMP_FASTQ_FILE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import subprocess\n",
    "fqFile = TEMP_FASTQ_FILE\n",
    "\n",
    "#specifying a list of parameters to run bowtie with\n",
    "#each tuple contains\n",
    "# *the mismatch threshold below which a site is considered a potential off-target (higher is more stringent)\n",
    "# *the number of sites allowed (1 is minimum since each sgRNA should have one true site in genome)\n",
    "# *the genome index against which to align the sgRNA sequences; these can be custom built to only consider sites near TSSs\n",
    "# *a name for the bowtie run to create appropriately named output files\n",
    "\n",
    "#iGEM\n",
    "#alignmentList = [(39,1,'~/indices/hg19.ensemblTSSflank500b','39_nearTSS'),\n",
    "#                (31,1,'~/indices/hg19.ensemblTSSflank500b','31_nearTSS'),\n",
    "#                (21,1,'~/indices/hg19.maskChrMandPAR','21_genome'),\n",
    "#                (31,2,'~/indices/hg19.ensemblTSSflank500b','31_2_nearTSS'),\n",
    "#                (31,3,'~/indices/hg19.ensemblTSSflank500b','31_3_nearTSS')]\n",
    "\n",
    "#iGEM   No alignments in hg19_maskChrMandPAR that need to be fixed. Need to generate hg19_maskChrMandPAR\n",
    "\n",
    "alignmentList = [(39,1,'../../../../data/large_data_files/indices/hg19.ensemblTSSflank500b','39_nearTSS'),\n",
    "                (31,1,'../../../../data/large_data_files/indices/hg19.ensemblTSSflank500b','31_nearTSS'),\n",
    "                #(21,1,'/data/large_data_files/indices/hg19_maskChrMandPAR','21_genome'),\n",
    "                (21,1,'../../../../data/large_data_files/indices/hg19','21_genome'),\n",
    "                (31,2,'../../../../data/large_data_files/indices/hg19.ensemblTSSflank500b','31_2_nearTSS'),\n",
    "                (31,3,'../../../../data/large_data_files/indices/hg19.ensemblTSSflank500b','31_3_nearTSS')]\n",
    "\n",
    "alignmentColumns = []\n",
    "for btThreshold, mflag, bowtieIndex, runname in alignmentList:\n",
    "\n",
    "    alignedFile = 'bowtie_output/' + runname + '_aligned.txt'\n",
    "    unalignedFile = 'bowtie_output/' + runname + '_unaligned.fq'\n",
    "    maxFile = 'bowtie_output/' + runname + '_max.fq'\n",
    "    \n",
    "    bowtieString = 'bowtie -n 3 -l 15 -e '+str(btThreshold)+' -m ' + str(mflag) + ' --nomaqround -a --tryhard -p 16 --chunkmbs 256 ' + bowtieIndex + ' --suppress 5,6,7 --un ' + unalignedFile + ' --max ' + maxFile + ' '+ ' -q '+fqFile+' '+ alignedFile\n",
    "    print (bowtieString)\n",
    "    print (subprocess.call(bowtieString, shell=True))\n",
    "\n",
    "    #parse through the file of sgRNAs that exceeded \"m\", the maximum allowable alignments, and mark \"True\" any that are found\n",
    "    with open(maxFile) as infile:\n",
    "        sgsAligning = set()\n",
    "        for i, line in enumerate(infile):\n",
    "            if i%4 == 0: #id line\n",
    "                sgsAligning.add(line.strip()[1:])\n",
    "\n",
    "    alignmentColumns.append(libraryTable_new.apply(lambda row: row.name in sgsAligning, axis=1))\n",
    "#iGEM zip is an object in python3    \n",
    "#collate results into a table, and flip the boolean values to yield the sgRNAs that passed filter as True\n",
    "alignmentTable = pd.concat(alignmentColumns,axis=1, keys=list(zip(*alignmentList))[3]).ne(True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pick the top sgRNAs for a library, given predicted activity scores and off-target filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#combine all generated data into one master table\n",
    "predictedScores_new.name = 'predicted score'\n",
    "v2Table = pd.concat((libraryTable_new, predictedScores_new, alignmentTable, sgInfoTable_new), axis=1, keys=['library table v2', 'predicted score', 'off-target filters', 'sgRNA info'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "#for our pCRISPRi/a-v2 vector, we append flanking sequences to each sgRNA sequence for cloning and require the oligo to contain\n",
    "#exactly 1 BstXI and BlpI site each for cloning, and exactly 0 SbfI sites for sequencing sample preparation\n",
    "restrictionSites = {re.compile('CCA......TGG'):1,\n",
    "                   re.compile('GCT.AGC'):1,\n",
    "                   re.compile('CCTGCAGG'):0}\n",
    "\n",
    "def matchREsites(sequence, REdict):\n",
    "    seq = sequence.upper()\n",
    "#iGEM in python 3 dict.iteritems is not present. So replace with dict.items        \n",
    "#    for resite, numMatchesExpected in restrictionSites.iteritems():\n",
    "    for resite, numMatchesExpected in restrictionSites.items():\n",
    "        if len(resite.findall(seq)) != numMatchesExpected:\n",
    "            return False\n",
    "        \n",
    "    return True\n",
    "\n",
    "def checkOverlaps(leftPosition, acceptedLeftPositions, nonoverlapMin):\n",
    "    for pos in acceptedLeftPositions:\n",
    "        if abs(pos - leftPosition) < nonoverlapMin:\n",
    "            return False\n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#flanking sequences\n",
    "upstreamConstant = 'CCACCTTGTTG'\n",
    "downstreamConstant = 'GTTTAAGAGCTAAGCTG'\n",
    "\n",
    "#minimum overlap between two sgRNAs targeting the same TSS\n",
    "nonoverlapMin = 3\n",
    "\n",
    "#number of sgRNAs to pick per gene/TSS\n",
    "sgRNAsToPick = 10\n",
    "\n",
    "#iGEM TODO need to enable 21_genome as well. temporarily disabled\n",
    "\n",
    "#list of off-target filter (or combinations of filters) levels, matching the names in the alignment table above\n",
    "offTargetLevels = [['31_nearTSS', '21_genome'],\n",
    "                  ['31_nearTSS'],\n",
    "                  ['21_genome'],\n",
    "                  ['31_2_nearTSS'],\n",
    "                  ['31_3_nearTSS']]\n",
    "\n",
    "#offTargetLevels = [ ['31_nearTSS'],\n",
    "#                  ['31_2_nearTSS'],\n",
    "#                  ['31_3_nearTSS']]\n",
    "\n",
    "\n",
    "#for each gene/TSS, go through each sgRNA in descending order of predicted score\n",
    "#if an sgRNA passes the restriction site, overlap, and off-target filters, accept it into the library\n",
    "#if the number of sgRNAs accepted is less than sgRNAsToPick, reduce off-target stringency by one and continue\n",
    "v2Groups = v2Table.groupby([('library table v2','gene'),('library table v2','transcripts')])\n",
    "newSgIds = []\n",
    "unfinishedTss = []\n",
    "for (gene, transcript), group in v2Groups:\n",
    "    geneSgIds = []\n",
    "    geneLeftPositions = []\n",
    "    empiricalSgIds = dict()\n",
    "    \n",
    "    stringency = 0\n",
    "#iGEM use sort_values instead of sort    \n",
    "    while len(geneSgIds) < sgRNAsToPick and stringency < len(offTargetLevels):\n",
    "        for sgId_v2, row in group.sort_values(('predicted score','predicted score'), ascending=False).iterrows():\n",
    "            oligoSeq = upstreamConstant + row[('library table v2','sequence')] + downstreamConstant\n",
    "            leftPos = row[('sgRNA info', 'position')] - (23 if row[('sgRNA info', 'strand')] == '-' else 0)\n",
    "            if len(geneSgIds) < sgRNAsToPick and row['off-target filters'].loc[offTargetLevels[stringency]].all() \\\n",
    "                and matchREsites(oligoSeq, restrictionSites) \\\n",
    "                and checkOverlaps(leftPos, geneLeftPositions, nonoverlapMin):\n",
    "                geneSgIds.append((sgId_v2,\n",
    "                                  gene,transcript,\n",
    "                                  row[('library table v2','sequence')], oligoSeq,\n",
    "                                  row[('predicted score','predicted score')], np.nan,\n",
    "                                 stringency))\n",
    "                geneLeftPositions.append(leftPos)\n",
    "                \n",
    "        stringency += 1\n",
    "            \n",
    "    if len(geneSgIds) < sgRNAsToPick:\n",
    "        unfinishedTss.append((gene, transcript)) #if the number of accepted sgRNAs is still less than sgRNAsToPick, discard gene\n",
    "    else:\n",
    "        newSgIds.extend(geneSgIds)\n",
    "        \n",
    "libraryTable_complete = pd.DataFrame(newSgIds, columns = ['sgID', 'gene', 'transcript','protospacer sequence', 'oligo sequence',\n",
    " 'predicted score', 'empirical score', 'off-target stringency']).set_index('sgID')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#number of sgRNAs accepted at each stringency level\n",
    "#iGEM newLibraryTable is not defined\n",
    "#newLibraryTable.groupby('off-target stringency').agg(len).iloc[:,0]\n",
    "libraryTable_complete.groupby('off-target stringency').agg(len).iloc[:,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#number of TSSs with fewer than required number of sgRNAs (and thus not included in the library)\n",
    "print (len(unfinishedTss))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#iGEM TODO Do not run this and beyond this cell. It need to be fixed after figuring out missing keys and semantics.\n",
    "\n",
    "#Note that empirical information from previous screens can be included as well--for example:\n",
    "geneToDisc = maxDiscriminantTable['best score'].groupby(level=0).agg(max).to_dict()\n",
    "thresh = 7\n",
    "empiricalBonus = .2\n",
    "\n",
    "upstreamConstant = 'CCACCTTGTTG'\n",
    "downstreamConstant = 'GTTTAAGAGCTAAGCTG'\n",
    "\n",
    "nonoverlapMin = 3\n",
    "\n",
    "sgRNAsToPick = 10\n",
    "\n",
    "offTargetLevels = [['31_nearTSS', '21_genome'],\n",
    "                  ['31_nearTSS'],\n",
    "                  ['21_genome'],\n",
    "                  ['31_2_nearTSS'],\n",
    "                  ['31_3_nearTSS']]\n",
    "offTargetLevels_v1 = [[s + '_v1' for s in l] for l in offTargetLevels]\n",
    "#iGEM comment vGroups and v1Table to make other code work. Need to figure out how tow bring these may be from v1 version \n",
    "#v1Groups = v1Table.groupby([('relative position','gene'),('relative position','transcript')])\n",
    "v2Groups = v2Table.groupby([('library table v2','gene'),('library table v2','transcripts')])\n",
    "\n",
    "newSgIds = []\n",
    "unfinishedTss = []\n",
    "for (gene, transcript), group in v2Groups:\n",
    "    geneSgIds = []\n",
    "    geneLeftPositions = []\n",
    "    empiricalSgIds = dict()\n",
    "    \n",
    "    stringency = 0\n",
    "    \n",
    "    while len(geneSgIds) < sgRNAsToPick and stringency < len(offTargetLevels):\n",
    "        \n",
    "        if gene in geneToDisc and geneToDisc[gene] >= thresh and (gene, transcript) in v1Groups.groups:\n",
    "\n",
    "            for sgId_v1, row in v1Groups.get_group((gene, transcript)).sort(('Empirical activity score','Empirical activity score'),ascending=False).iterrows():\n",
    "                oligoSeq = upstreamConstant + row[('library table v2','sequence')] + downstreamConstant\n",
    "                leftPos = row[('sgRNA info', 'position')] - (23 if row[('sgRNA info', 'strand')] == '-' else 0)\n",
    "                if len(geneSgIds) < sgRNAsToPick and min(abs(row.loc['relative position'].iloc[2:])) < 5000 \\\n",
    "                and row[('Empirical activity score','Empirical activity score')] >= .75 \\\n",
    "                and row['off-target filters'].loc[offTargetLevels_v1[stringency]].all() \\\n",
    "                and matchREsites(oligoSeq, restrictionSites) \\\n",
    "                and checkOverlaps(leftPos, geneLeftPositions, nonoverlapMin):\n",
    "                    if len(geneSgIds) < 2:\n",
    "                        geneSgIds.append((row[('library table v2','sgId_v2')],\n",
    "                                          gene,transcript,\n",
    "                                          row[('library table v2','sequence')], oligoSeq,\n",
    "                                          np.nan,row[('Empirical activity score','Empirical activity score')],\n",
    "                                         stringency))\n",
    "                        geneLeftPositions.append(leftPos)\n",
    "\n",
    "                    empiricalSgIds[row[('library table v2','sgId_v2')]] = row[('Empirical activity score','Empirical activity score')]\n",
    "\n",
    "        adjustedScores = group.apply(lambda row: row[('predicted score','CRISPRiv2 predicted score')] + empiricalBonus if row.name in empiricalSgIds else row[('predicted score','CRISPRiv2 predicted score')], axis=1)\n",
    "        adjustedScores.name = ('adjusted score','')\n",
    "        for sgId_v2, row in pd.concat((group,adjustedScores),axis=1).sort(('adjusted score',''), ascending=False).iterrows():\n",
    "            oligoSeq = upstreamConstant + row[('library table v2','sequence')] + downstreamConstant\n",
    "            leftPos = row[('sgRNA info', 'position')] - (23 if row[('sgRNA info', 'strand')] == '-' else 0)\n",
    "            if len(geneSgIds) < sgRNAsToPick and row['off-target filters'].loc[offTargetLevels[stringency]].all() \\\n",
    "                and matchREsites(oligoSeq, restrictionSites) \\\n",
    "                and checkOverlaps(leftPos, geneLeftPositions, nonoverlapMin):\n",
    "                geneSgIds.append((sgId_v2,\n",
    "                                  gene,transcript,\n",
    "                                  row[('library table v2','sequence')], oligoSeq,\n",
    "                                  row[('predicted score','CRISPRiv2 predicted score')], empiricalSgIds[sgId_v2] if sgId_v2 in empiricalSgIds else np.nan,\n",
    "                                 stringency))\n",
    "                geneLeftPositions.append(leftPos)\n",
    "                \n",
    "        stringency += 1\n",
    "            \n",
    "    if len(geneSgIds) < sgRNAsToPick:\n",
    "        unfinishedTss.append((gene, transcript))\n",
    "    else:\n",
    "        newSgIds.extend(geneSgIds)\n",
    "\n",
    "    \n",
    "libraryTable_complete = pd.DataFrame(newSgIds, columns = ['sgID', 'gene', 'transcript','protospacer sequence', 'oligo sequence',\n",
    " 'predicted score', 'empirical score', 'off-target stringency']).set_index('sgID')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Design negative controls matching the base composition of the library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calcluate the base frequency at each position of the sgRNA, then generate random sequences weighted by this frequency\n",
    "def getBaseFrequencies(libraryTable, baseConversion = {'G':0, 'C':1, 'T':2, 'A':3}):\n",
    "    baseArray = np.zeros((len(libraryTable),20))\n",
    "\n",
    "    for i, (index, seq) in enumerate(libraryTable['protospacer sequence'].iteritems()):\n",
    "        for j, char in enumerate(seq.upper()):\n",
    "            baseArray[i,j] = baseConversion[char]\n",
    "\n",
    "    baseTable = pd.DataFrame(baseArray, index = libraryTable.index)\n",
    "    \n",
    "    baseFrequencies = baseTable.apply(lambda col: col.groupby(col).agg(len)).fillna(0) / len(baseTable)\n",
    "    baseFrequencies.index = ['G','C','T','A']\n",
    "    \n",
    "    baseCumulativeFrequencies = baseFrequencies.copy()\n",
    "    baseCumulativeFrequencies.loc['C'] = baseFrequencies.loc['G'] + baseFrequencies.loc['C']\n",
    "    baseCumulativeFrequencies.loc['T'] = baseFrequencies.loc['G'] + baseFrequencies.loc['C'] + baseFrequencies.loc['T']\n",
    "    baseCumulativeFrequencies.loc['A'] = baseFrequencies.loc['G'] + baseFrequencies.loc['C'] + baseFrequencies.loc['T'] + baseFrequencies.loc['A']\n",
    "\n",
    "    return baseFrequencies, baseCumulativeFrequencies\n",
    "\n",
    "def generateRandomSequence(baseCumulativeFrequencies):\n",
    "    randArray = np.random.random(baseCumulativeFrequencies.shape[1])\n",
    "    \n",
    "    seq = []\n",
    "    for i, col in baseCumulativeFrequencies.iteritems():\n",
    "        for base, freq in col.iteritems():\n",
    "            if randArray[i] < freq:\n",
    "                seq.append(base)\n",
    "                break\n",
    "                \n",
    "    return ''.join(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import pdb\n",
    "#pdb.set_trace()\n",
    "\n",
    "baseCumulativeFrequencies = getBaseFrequencies(libraryTable_complete)[1]\n",
    "negList = []\n",
    "#for i in range(30000):\n",
    "#    negList.append(generateRandomSequence(baseCumulativeFrequencies))\n",
    "#negTable = pd.DataFrame(negList, index=['non-targeting_' + str(i) for i in range(30000)], columns = ['sequence'])\n",
    "\n",
    "numberToGenerate = 10 #can generate many more; some will be filtered out by off-targets, and you can always select an arbitrary subset for inclusion into the library\n",
    "for i in range(numberToGenerate):\n",
    "    negList.append(generateRandomSequence(baseCumulativeFrequencies))\n",
    "negTable = pd.DataFrame(negList, index=['non-targeting_' + str(i) for i in range(numberToGenerate)], columns = ['sequence'])\n",
    "\n",
    "\n",
    "outputTempBowtieFastq(negTable, TEMP_FASTQ_FILE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#similar to targeting sgRNA off-target scoring, but looking for sgRNAs with 0 alignments\n",
    "fqFile = TEMP_FASTQ_FILE\n",
    "\n",
    "alignmentList = [(31,1,'~/indices/hg19.ensemblTSSflank500b','31_nearTSS_negs'),\n",
    "                (21,1,'~/indices/hg19.maskChrMandPAR','21_genome_negs')]\n",
    "\n",
    "alignmentColumns = []\n",
    "for btThreshold, mflag, bowtieIndex, runname in alignmentList:\n",
    "\n",
    "    alignedFile = 'bowtie_output/' + runname + '_aligned.txt'\n",
    "    unalignedFile = 'bowtie_output//' + runname + '_unaligned.fq'\n",
    "    maxFile = 'bowtie_output/' + runname + '_max.fq'\n",
    "    \n",
    "    bowtieString = 'bowtie -n 3 -l 15 -e '+str(btThreshold)+' -m ' + str(mflag) + ' --nomaqround -a --tryhard -p 16 --chunkmbs 256 ' + bowtieIndex + ' --suppress 5,6,7 --un ' + unalignedFile + ' --max ' + maxFile + ' '+ ' -q '+fqFile+' '+ alignedFile\n",
    "    print bowtieString\n",
    "    print subprocess.call(bowtieString, shell=True)\n",
    "\n",
    "    #read unaligned file for negs, and then don't flip boolean of alignmentTable\n",
    "    with open(unalignedFile) as infile:\n",
    "        sgsAligning = set()\n",
    "        for i, line in enumerate(infile):\n",
    "            if i%4 == 0: #id line\n",
    "                sgsAligning.add(line.strip()[1:])\n",
    "\n",
    "    alignmentColumns.append(negTable.apply(lambda row: row.name in sgsAligning, axis=1))\n",
    "    \n",
    "alignmentTable = pd.concat(alignmentColumns,axis=1, keys=zip(*alignmentList)[3])\n",
    "alignmentTable.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "acceptedNegList = []\n",
    "negCount = 0\n",
    "for i, (name, row) in enumerate(pd.concat((negTable,alignmentTable),axis=1, keys=['seq','alignment']).iterrows()):\n",
    "    oligo = upstreamConstant + row['seq','sequence'] + downstreamConstant\n",
    "    if row['alignment'].all() and matchREsites(oligo, restrictionSites):\n",
    "        acceptedNegList.append(('non-targeting_%05d' % negCount, 'negative_control', 'na', row['seq','sequence'], oligo, 0))\n",
    "        negCount += 1\n",
    "        \n",
    "acceptedNegs = pd.DataFrame(acceptedNegList, columns = ['sgId', 'gene', 'transcript', 'protospacer sequence', 'oligo sequence', 'off-target stringency']).set_index('sgId')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finalizing library design"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* divide genes into sublibrary groups (if required)\n",
    "* assign negative control sgRNAs to sublibrary groups; ~1-2% of the number of sgRNAs in the library is a good rule-of-thumb\n",
    "* append PCR adapter sequences (~18bp) to each end of the oligo sequences to enable amplification of the oligo pool; each sublibary should have an orthogonal sequence so they can be cloned separately"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "#import pdb\n",
    "#pdb.set_trace()\n",
    "print (sys.version)\n",
    "#transformedParamHeader.head()\n",
    "#transformedParamHeader=transformedParamsTrainHead\n",
    "#paramTable_trainingGuides.describe()\n",
    "from sklearn import linear_model\n",
    "reg = linear_model.LinearRegression()\n",
    "reg.fit([[0, 0], [1, 1], [2, 2]], [0, 1, 2])\n",
    "#LinearRegression()\n",
    "reg.coef_\n",
    "reg.intercept_"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
